### PR TITLE
update mysql fix tool to fit resourceQuota schema

### PR DESCRIPTION
**Pull request by YinYangOfDao**
support updating DB in deploy.py, support ratio-based auto-resource-assignment

### CODE REVIEW

File: src/ClusterBootstrap/deploy.py [hunk 1]
```diff
@@ -3408,6 +3408,12 @@ def run_command(args, command, nargs, parser):
         if nargs[0] == "config":
             update_config_nodes()
 
+    elif command == "updatemysqlvc":
+        from MySQLDataHandler import update_db
+        with open("deploy/RestfulAPI/config.yaml") as f:
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/deploy.py]: Why loading `config.yaml` again? All configurations should already be in `config.yaml`?

Comment (YinYangOfDao) [reply src/ClusterBootstrap/deploy.py]: no, it's different yaml file. this MysqlDatahandler.py reads the config in RestfulAPI container, as in the code. deploy.py instead only reads ClusterBootstrap/config.yaml and cluster.yaml, if we don't read the one in the code, there would be key error

File: src/utils/MySQLDataHandler.py [hunk 2]
```diff
@@ -247,6 +353,8 @@ def CreateTable(self):
             # when the VC has vm of same GPU type but different VMsizes, e.g., when VC has Standard_NC6s_v3 and Standard_NC12s_v3 both?
             # impossible since there's no way to do it with current config mechanism
 
+            vc_table_exist = table_exists(self.conn, self.vctablename)
```
Comment (Anbang-Hu) [inline src/utils/MySQLDataHandler.py]: Why not make `table_exists` a class method?

File: src/utils/MySQLDataHandler.py [hunk 3]
```diff
@@ -260,17 +368,19 @@ def CreateTable(self):
...
-            sql += default_vc_value
-
             cursor = self.conn.cursor()
             cursor.execute(sql)
             self.conn.commit()
             cursor.close()
 
+            if not vc_table_exist:
+                defalult_vc_rows = insert_vc_sqls(config)
+                execute_mysql(self.conn, defalult_vc_rows)
```
Comment (Anbang-Hu) [inline src/utils/MySQLDataHandler.py]: There is a class method called `AddVC`

File: src/utils/MySQLDataHandler.py [hunk 4]
```diff
@@ -43,10 +43,102 @@ def base64decode(str_val):
...
+    return sqls
+
+
+def column_exists(conn, tablename, columnname):
+    sql = "SHOW COLUMNS FROM `{}` LIKE '{}';".format(tablename, columnname)
+    cursor = conn.cursor()
+    cursor.execute(sql)
+    result = cursor.fetchone()
+    print(result)
+    return not result is None
```
Comment (Anbang-Hu) [inline src/utils/MySQLDataHandler.py]: `result is not None`

File: src/utils/MySQLDataHandler.py [hunk 5]
```diff
@@ -43,10 +43,102 @@ def base64decode(str_val):
...
+    total = 0
+    for entry in myresult:
+        total += sum(eval(entry[4]).values())
+    ratio = {}
+    for entry in myresult:
+        ratio[str(entry[2])] = sum(eval(entry[4]).values()) / total
+    return ratio
+
+
+def update_db(config):
```
Comment (Anbang-Hu) [inline src/utils/MySQLDataHandler.py]: For update VC table, it's better to use `UpdateVC`

File: src/utils/MySQLDataHandler.py [hunk 6]
```diff
@@ -330,6 +274,23 @@ def CreateTable(self):
             self.conn.commit()
             cursor.close()
 
+    @record
+    def column_exists(self, tablename, columnname):
+        sql = "SHOW COLUMNS FROM `{}` LIKE '{}';".format(tablename, columnname)
+        cursor = self.conn.cursor()
+        cursor.execute(sql)
+        result = cursor.fetchone()
+        print(result)
```
Comment (Anbang-Hu) [inline src/utils/MySQLDataHandler.py]: Remove print. If you really want to print, use `logger`

Comment (Anbang-Hu) [reply src/utils/MySQLDataHandler.py]: Same for line 292, 367, 373

File: src/utils/MySQLDataHandler.py [hunk 7]
```diff
@@ -393,13 +354,25 @@ def UpdateStorage(self, vcName, url, storageType, metadata, defaultMountPath):
             logger.error('Exception: %s', str(e))
             return False
 
-
     @record
```
Comment (Anbang-Hu) [inline src/utils/MySQLDataHandler.py]: Please move `@record` annotation to where it belonged - above `AddVC`

File: src/ClusterBootstrap/step_by_step.sh [hunk 8]
```diff
@@ -21,7 +21,6 @@ fi
 ./deploy.py labelsku
 ./deploy.py webui
 ./deploy.py docker push restfulapi
-./deploy.py docker push webui
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/step_by_step.sh]: Why removing this line?

File: src/utils/MySQLDataHandler.py [hunk 9]
```diff
@@ -393,13 +354,25 @@ def UpdateStorage(self, vcName, url, storageType, metadata, defaultMountPath):
...
-    def AddVC(self, vcName, quota, metadata):
+    def init_vc_sqls(self, config):
+        ratio_dict = {config['defalt_virtual_cluster_name']: 1.0}
+        if "vc_resource_ratio" in config:
+            ratio_dict = config["vc_resource_ratio"]
+        quota, metadata, res_quota, res_meta = vc_value_str(config, ratio_dict)
+        if quota == "":
+            return
+        sqls = []
```
Comment (Anbang-Hu) [inline src/utils/MySQLDataHandler.py]: What is `sqls = []` for?

File: src/utils/MySQLDataHandler.py [hunk 10]
```diff
@@ -393,13 +354,25 @@ def UpdateStorage(self, vcName, url, storageType, metadata, defaultMountPath):
...
+        for i, (vc, vc_res_quota) in enumerate(res_quota.items()):
```
Comment (Anbang-Hu) [inline src/utils/MySQLDataHandler.py]: `i` is not used. The for loop can be simplified to `for vc, vc_res_quota in res_quota.items():`

File: src/utils/vc_quota.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    res_meta_dict = {}
+    for r_type in ["cpu", "memory", "gpu", "gpu_memory"]:
+        tmp_res_meta = {}
+        for sku in worker_sku_cnt:
+            sku_name_in_map = sku if sku in sku_mapping else ""
+            pernode_cnt = sku_mapping.get(sku_name_in_map, {}).get(r_type, 0)
+            if "memory" in r_type:
+                pernode_cnt = '{}Gi'.format(pernode_cnt)
+            tmp_res_meta[sku_name_in_map] = {"per_node": pernode_cnt}
+            if r_type == "gpu":
```
Comment (Anbang-Hu) [inline src/utils/vc_quota.py]: Need to add `unschedulable_ratio: 0.9` for CPU and memory. Check dev1's `vc` table.

File: src/ClusterBootstrap/az_tools.py [hunk 12]
```diff
@@ -1212,9 +1212,6 @@ def run_command(args, command, nargs, parser):
 
     config = update_config(config)
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/az_tools.py]: Can you revert the change in `az_tools.py`? It was checked in in #875

File: src/utils/MySQLDataHandler.py (hunk 7 above)
Comment (Anbang-Hu) [reply src/utils/MySQLDataHandler.py]: Please remove `@record` here. `init_vc_sqls` internally calls `AddVC`. It is not a direct DB call itself.
